### PR TITLE
User registration feature

### PR DESCRIPTION
**Pull request by AbegaM**
### Modifications 

1. Added a middleware named `processRequest` to remove some fields from the `/api/tables/_users` API call
2. Added a middleware named `processResponse` to redirect an API call to the user registration endpoint

### CODE REVIEW

File: src/controllers/rows.js [hunk 1]
```diff
@@ -1,11 +1,11 @@
 const db = require('../db/index');
```
Comment (IanMayo) [inline src/controllers/rows.js:1]: @AbegaM - there are lots of cosmetic (trailing comma) changes in this file, which make it harder to review. Plus, the next person's PR will probably try to fix them - more unnecessary changes.

We need to understand why some blocks of working code have been commented out or changes made.  No problem if this is still WIP though.

Comment (AbegaM) [reply src/controllers/rows.js:1]: There is some inconsistency in `Soul` for example, some files are using `double quotes` and other files are using `single quotes` and there are some trailing commas in some files but not in the others, and i have configured my formatter to remove trailing commas and that's why the changes are here, but regarding the lines i commented i can remove the commented code

**Comment by TahaKhanAbdalli** [comment]
`Auth mode enabled and not enabled`.
I have tested this by creating the `.env` file and assigning `AUTH = true`, which enables the `Auth` mode. Subsequently, it creates the tables in `foobar.db`.
However, if the `Auth` mode is not enabled, it creates the database with the name `foobar.db` but does not create the tables. Then, when I attempt to create tables manually and create a user using the endpoint `/api/tables/_users/rows`, it successfully creates the user but returns the following error message:
`Please restart soul so a default role can be created`.
So, if the `Auth` is disabled, why is this error message showing?

**Comment by AbegaM** [comment]
> `Auth mode enabled and not enabled`. I have tested this by creating the `.env` file and assigning `AUTH = true`, which enables the `Auth` mode. Subsequently, it creates the tables in `foobar.db`. However, if the `Auth` mode is not enabled, it creates the database with the name `foobar.db` but does not create the tables. Then, when I attempt to create tables manually and create a user using the endpoint `/api/tables/_users/rows`, it successfully creates the user but returns the following error message: `Please restart soul so a default role can be created`. So, if the `Auth` is disabled, why is this error message showing?

Hello @TahaKhanAbdalli, if the `AUTH` mode is disabled, there is no need to create users. The registration process will be skipped, and there will be no need to create roles. In this case, we have planned to make the `_users` table name reserved, meaning it will not be allowed to create a table with this name in the DB, so you only need to test the authentication feature by setting `AUTH` to `true`

**Comment by TahaKhanAbdalli** [comment]
Okay, that should resolve the error. But in my opinion, there should be an additional check which confirms whether the `AUTH` is enabled or not.

**Comment by AbegaM** [comment]
> Okay, that should resolve the error. But in my opinion, there should be an additional check which confirms whether the `AUTH` is enabled or not.

Sure, when a user tries to access the reserved resources like `_roles`, `_roles_permissions`, `_users` and `_users_roles` while `AUTH` is set to `false` we can throw an error saying, `You can't access this endpoint while AUTH is set to false`

**Comment by IanMayo** [comment]
But @AbegaM - this looks like a legitimate use of the app.  It may be a bit niche, but I see the logic in being able to do user management with auth (temporarily) switched off, then switch auth back on.

> Please restart soul so a default role can be created

This warning suggest code is going down a route it shouldn't, if auth is off.

**Comment by AbegaM** [comment]
> But @AbegaM - this looks like a legitimate use of the app. It may be a bit niche, but I see the logic in being able to do user management with auth (temporarily) switched off, then switch auth back on.
> 
> > Please restart soul so a default role can be created
> 
> This warning suggest code is going down a route it shouldn't, if auth is off.

Note that this error message is not shown when `AUTH` is disabled, if `AUTH` is disabled there is no reason for the user to use the `/_users` endpoint, The error is thrown when the code fails to get the `default` role in the DB

**Comment by AbegaM** [comment]
@IanMayo and we have decided to make the default table names `reserved` so if the user tries to set the `AUTH` to `false` and then tries to create a table by using the reserved table names like `_users` they will get an error

**Comment by IanMayo** [comment]
> @IanMayo and we have decided to make the default table names `reserved` so if the user tries to set the `AUTH` to `false` and then tries to create a table by using the reserved table names like `_users` they will get an error

Sure. Tables with reserved names cannot be created.  But, I see `running with auth disabled` as being like `running as super-user` - able to change (almost) any thing.

**Comment by AbegaM** [comment]
@IanMayo The issue you mentioned about Souls `default` endpoints being accessed while `AUTH` is set to `false` is fixed now, i have added a check that throws an error when a user tries to access the `default` endpoints when `AUTH` is set to false

**Comment by AbegaM** [comment]
@thevahidal The comments in this PR are fixed, please let us know if you have additional comments

File: src/routes/rows.js [hunk 2]
```diff
@@ -3,37 +3,42 @@ const express = require('express');
...
   controllers.updateRowInTableByPK,
-  broadcast
+  broadcast,
 );
 router.delete(
   '/:name/rows/:pks',
   validator(schema.deleteRowInTableByPK),
   controllers.deleteRowInTableByPK,
-  broadcast
+  broadcast,
```
Comment (thevahidal) [inline src/routes/rows.js:41]: Shouldn't we use the process request and response here?

**Review by thevahidal — CHANGES_REQUESTED** [review]
Thanks @AbegaM, 
I added some comments here.

File: src/routes/rows.js (hunk 2 above)
Comment (AbegaM) [reply src/routes/rows.js:41]: Those middleware functions are added in the next PR

File: src/controllers/auth.js [hunk 3]
```diff
@@ -133,4 +134,66 @@ const updateSuperuser = async (fields) => {
...
+      tableName: '_roles',
+      whereString: 'WHERE name=?',
+      whereStringValues: ['default'],
+    });
+
+    if (defaultRole.length <= 0) {
+      return res.status(500).send({
+        message: 'Please restart soul so a default role can be created',
+      });
+    }
```
[the remark this answers is not in the document]
Comment (AbegaM) [reply src/controllers/auth.js:194]: @thevahidal Thank you for the review, All of the new comments are fixed excluding this one ☝️

[the remark this answers is not in the document]
Comment (thevahidal) [reply src/controllers/auth.js:194]: Great job man, I approve this PR too.

**Review by thevahidal — APPROVED** [review]
